File: Adapters/Vungle/ISVungleAdapter/src/main/java/com/ironsource/adapters/vungle/VungleInterstitialAdapter.java
package com.ironsource.adapters.vungle;

import com.ironsource.environment.ContextProvider;
import com.ironsource.mediationsdk.logger.IronLog;
import com.ironsource.mediationsdk.logger.IronSourceError;
import com.ironsource.mediationsdk.sdk.InterstitialSmashListener;
import com.ironsource.mediationsdk.utils.ErrorBuilder;
import com.ironsource.mediationsdk.utils.IronSourceConstants;
import com.vungle.ads.AdConfig;
import com.vungle.ads.BaseAd;
import com.vungle.ads.InterstitialAd;
import com.vungle.ads.InterstitialAdListener;
import com.vungle.ads.VungleException;

final class VungleInterstitialAdapter implements InterstitialAdListener {

    private InterstitialSmashListener mListener;
    private InterstitialAd mInterstitialAd;

    VungleInterstitialAdapter(String placementId, AdConfig adConfig, InterstitialSmashListener listener) {
        this.mListener = listener;

        if (adConfig == null) {
            adConfig = new AdConfig();
        }

        mInterstitialAd = new InterstitialAd(ContextProvider.getInstance().getApplicationContext(), placementId, adConfig);
        mInterstitialAd.setAdListener(this);
    }

    void load() {
        mInterstitialAd.load(null);
    }

    public void loadWithBid(String serverData) {
        mInterstitialAd.load(serverData);
    }

    public boolean canPlayAd() {
        return mInterstitialAd.canPlayAd();
    }

    public void play() {
        mInterstitialAd.play();
    }

    public void destroy() {
        mListener = null;
        mInterstitialAd = null;
    }

    @Override
    public void onAdLoaded(BaseAd baseAd) {
        IronLog.ADAPTER_CALLBACK.verbose("placementId = " + baseAd.getPlacementId());

        if (mListener == null) {
            IronLog.INTERNAL.verbose("listener is null");
            return;
        }

        mListener.onInterstitialAdReady();
    }

    @Override
    public void onAdStart(BaseAd baseAd) {
        IronLog.ADAPTER_CALLBACK.verbose("placementId = " + baseAd.getPlacementId());

        if (mListener == null) {
            IronLog.INTERNAL.verbose("listener is null");
            return;
        }

        mListener.onInterstitialAdShowSucceeded();
    }

    @Override
    public void onAdImpression(BaseAd baseAd) {
        IronLog.ADAPTER_CALLBACK.verbose("placementId = " + baseAd.getPlacementId());

        if (mListener == null) {
            IronLog.INTERNAL.verbose("listener is null");
            return;
        }

        mListener.onInterstitialAdOpened();
    }

    @Override
    public void onAdClicked(BaseAd baseAd) {
        IronLog.ADAPTER_CALLBACK.verbose("placementId = " + baseAd.getPlacementId());

        if (mListener == null) {
            IronLog.INTERNAL.verbose("listener is null");
            return;
        }

        mListener.onInterstitialAdClicked();
    }

    @Override
    public void onAdEnd(BaseAd baseAd) {
        IronLog.ADAPTER_CALLBACK.verbose("placementId = " + baseAd.getPlacementId());

        if (mListener == null) {
            IronLog.INTERNAL.verbose("listener is null");
            return;
        }

        mListener.onInterstitialAdClosed();
    }

    @Override
    public void onAdFailedToPlay(BaseAd baseAd, VungleException exception) {
        IronLog.ADAPTER_CALLBACK.verbose("onAdFailedToPlay placementId = " + baseAd.getPlacementId() + ", exception = " + exception);

        if (mListener == null) {
            IronLog.INTERNAL.verbose("listener is null");
            return;
        }

        String errorMessage = " reason = " + exception.getLocalizedMessage() + " errorCode = " + exception.getExceptionCode();
        mListener.onInterstitialAdShowFailed(ErrorBuilder.buildShowFailedError(IronSourceConstants.INTERSTITIAL_AD_UNIT, errorMessage));
    }

    @Override
    public void onAdFailedToLoad(BaseAd baseAd, VungleException exception) {
        IronLog.ADAPTER_CALLBACK.verbose("onAdFailedToLoad placementId = " + baseAd.getPlacementId() + ", exception = " + exception);

        if (mListener == null) {
            IronLog.INTERNAL.verbose("listener is null");
            return;
        }

        IronSourceError error;
        if (exception.getExceptionCode() == VungleException.NO_SERVE) {
            error = new IronSourceError(IronSourceError.ERROR_IS_LOAD_NO_FILL, exception.getLocalizedMessage());
        } else {
            error = ErrorBuilder.buildLoadFailedError(exception.getLocalizedMessage());
        }

        mListener.onInterstitialAdLoadFailed(error);
    }

    @Override
    public void onAdLeftApplication(BaseAd baseAd) {
        // no-op
    }

}

File: Adapters/Vungle/ISVungleAdapter/src/main/java/com/ironsource/adapters/vungle/VungleInterstitialAdapter.kt
package com.ironsource.adapters.vungle

import com.ironsource.environment.ContextProvider
import com.ironsource.mediationsdk.logger.IronLog
import com.ironsource.mediationsdk.logger.IronSourceError
import com.ironsource.mediationsdk.sdk.InterstitialSmashListener
import com.ironsource.mediationsdk.utils.ErrorBuilder
import com.ironsource.mediationsdk.utils.IronSourceConstants
import com.vungle.ads.AdConfig
import com.vungle.ads.BaseAd
import com.vungle.ads.InterstitialAd
import com.vungle.ads.InterstitialAdListener
import com.vungle.ads.VungleError

internal class VungleInterstitialAdapter(
    placementId: String,
    adConfig: AdConfig,
    private val listener: InterstitialSmashListener?
) : InterstitialAdListener {
    private var mInterstitialAd: InterstitialAd?

    init {
        mInterstitialAd = InterstitialAd(
            ContextProvider.getInstance().applicationContext,
            placementId,
            adConfig
        )
        mInterstitialAd?.adListener = this
    }

    fun loadWithBid(serverData: String?) {
        mInterstitialAd?.load(serverData)
    }

    fun canPlayAd(): Boolean {
        return mInterstitialAd?.canPlayAd() == true
    }

    fun play() {
        mInterstitialAd?.play()
    }

    fun destroy() {
        mInterstitialAd = null
    }

    override fun onAdLoaded(baseAd: BaseAd) {
        IronLog.ADAPTER_CALLBACK.verbose("placementId = ${baseAd.placementId}")
        if (listener == null) {
            IronLog.INTERNAL.verbose("listener is null")
            return
        }
        listener.onInterstitialAdReady()
    }

    override fun onAdStart(baseAd: BaseAd) {
        IronLog.ADAPTER_CALLBACK.verbose("placementId = ${baseAd.placementId}")
        if (listener == null) {
            IronLog.INTERNAL.verbose("listener is null")
            return
        }
        listener.onInterstitialAdShowSucceeded()
    }

    override fun onAdImpression(baseAd: BaseAd) {
        IronLog.ADAPTER_CALLBACK.verbose("placementId = ${baseAd.placementId}")
        if (listener == null) {
            IronLog.INTERNAL.verbose("listener is null")
            return
        }
        listener.onInterstitialAdOpened()
    }

    override fun onAdClicked(baseAd: BaseAd) {
        IronLog.ADAPTER_CALLBACK.verbose("placementId = ${baseAd.placementId}")
        if (listener == null) {
            IronLog.INTERNAL.verbose("listener is null")
            return
        }
        listener.onInterstitialAdClicked()
    }

    override fun onAdEnd(baseAd: BaseAd) {
        IronLog.ADAPTER_CALLBACK.verbose("placementId = ${baseAd.placementId}")
        if (listener == null) {
            IronLog.INTERNAL.verbose("listener is null")
            return
        }
        listener.onInterstitialAdClosed()
    }

    override fun onAdFailedToPlay(baseAd: BaseAd, adError: VungleError) {
        IronLog.ADAPTER_CALLBACK.verbose("onAdFailedToPlay placementId = ${baseAd.placementId}, error = $adError")
        if (listener == null) {
            IronLog.INTERNAL.verbose("listener is null")
            return
        }
        val errorMessage = " reason = " + adError.errorMessage + " errorCode = " + adError.code
        listener.onInterstitialAdShowFailed(
            ErrorBuilder.buildShowFailedError(
                IronSourceConstants.INTERSTITIAL_AD_UNIT,
                errorMessage
            )
        )
    }

    override fun onAdFailedToLoad(baseAd: BaseAd, adError: VungleError) {
        IronLog.ADAPTER_CALLBACK.verbose("onAdFailedToLoad placementId = ${baseAd.placementId}, error = $adError")
        if (listener == null) {
            IronLog.INTERNAL.verbose("listener is null")
            return
        }
        val error: IronSourceError = if (adError.code == VungleError.NO_SERVE) {
            IronSourceError(IronSourceError.ERROR_IS_LOAD_NO_FILL, adError.errorMessage)
        } else {
            ErrorBuilder.buildLoadFailedError(adError.errorMessage)
        }
        listener.onInterstitialAdLoadFailed(error)
    }

    override fun onAdLeftApplication(baseAd: BaseAd) {
        // no-op
    }
}
